src/hardware/sblaster.cpp

```diff
@@ -829,7 +829,20 @@ static void dsp_dma_callback(const DmaChannel* chan, const DmaEvent event)
 			// ready for playback. This is when we set the callback running to
 			// play the data.
 			sblaster->MaybeWakeUp();
-			callback_type.SetPerTick();
+
+			// If the DMA transfer is setup with a base count of fewer than
+			// three elements (which is four bytes given one 16-bit stereo frame
+			// held in an 8-bit DMA channel), then we know the software intends
+			// to overwrite the DMA content on the fly instead of pre-generating
+			// large chunks of DMA audio. In these cases we prefer the
+			// fine-grained per-frame callback. (The minus one is because DMA
+			// counts are in addition to one; so a base count of zero is one
+			// element).
+			constexpr auto MaxSingleFrameBaseCount = sizeof(int16_t) * 2 - 1;
+
+			(chan->base_count <= MaxSingleFrameBaseCount)
+			        ? callback_type.SetPerFrame()
+			        : callback_type.SetPerTick();
```

Comment:
I still don''t understand what the benefit of these per frame callbacks are.  Theoretically it should end up doing exactly the same thing since these callbacks can only run at most once per 1ms.

I'm fine leaving it for now though.  You've got both methods in there so it should be an easy swap if this runs into any problems in the future.

Reply:
It looks that way on the surface, but each PIC event is scheduled to run at some fractional amount of milliseconds in the future. 

The fractional value is a double, and the event is run at exactly at that ratio of cycle position within the allotted total cycles.

So for example, if you schedule a pic event with a value of 0.5ms, and if the system is running at fixed 3000 cycles/ms, then the event will trigger at exactly the 1500-cycle mark. So you can get events happening inside the 1ms cycle tick.

If you scheduled it for 0.1, it would trigger at the 300-cycle mark, followed by 2700 more cpu cycles after the event before the emulator's full 1ms round is done. 

So to play audio from a memory address that Crystal Dreams is writting to at 16 KHz, we need the SB reading from DMA using PIC events that are each scheduled  1/16000s in the future.

Let's say dosbox is running Crystal Dreams at a cycle count of 10000 cycles/ms. Our per frame events will fire after every 625 cycles elapses, which is exactly 16 times within the 1ms emulator round (where it's doing 10k cycles/ms). 

If you print the PIC full index you can see that fractional time count by. That's how I worked backward to land on this approach. 

This differs from the timer callback system that only gives you 1ms granularity, and those callbacks are only run after the entire set of cycles is complete.

Reply:
Nice explanation @interloper98 👍🏻

Reply:
Got it.  I was under the impression that each 1ms tick always finished in its entirety before any PIC events could be run.  I take it you've looked into this more than I have though :smile: